### ISSUE TITLE
rewrite_strat regression in 8.15

### ISSUE THREAD

**Issue by rlepigre** [coq/coq#15568]
#### Description of the problem

The following example demonstrates a problem with `rewrite_strat` which does not arise in 8.13 and 8.14. There seem to be some weird interplay with type class search: removing either or both of the `Reflexive` and `Transitive` instances also fixes the problem.

```coq
Require Import Coq.ZArith.BinInt.
Require Import Coq.Classes.RelationClasses.

Axiom mod_mod : forall a b : Z, Z.modulo (Z.modulo b a) a = Z.modulo b a.

#[global] Hint Rewrite Z.add_0_r mod_mod : bu.

Definition modEq (d:Z) := fun (a b :Z)=> Z.modulo a d = Z.modulo b d.

#[global] Instance modEqR (d:Z): Reflexive  (modEq d). Admitted.
#[global] Instance modEqT (d:Z): Transitive (modEq d). Admitted.

Goal forall d a e, Z.modulo (Z.modulo a d + 0) d = e.
intros.
Fail rewrite_strat try (bottomup (hints bu)). (* Should not fail. *)
(*
Error: Illegal application: 
The term "modEqT" of type "forall d : Z, Transitive (modEq d)"
cannot be applied to the terms
 "?d" : "Z"
 "((Zmodd d a + 0) mod d)%Z" : "Z"
 "(Zmodd d a mod d)%Z" : "Z"
 "(a mod d)%Z" : "Z"
 "reflexive_proper Z.modulo (Zmodd d a + 0)%Z (Zmodd d a) (Z.add_0_r (Zmodd d a)) d d (eq_proper_proxy d)"
   : "modEq ?d ((Zmodd d a + 0) mod d) (Zmodd d a mod d)"
 "Zmodd_mod d a" : "(Zmodd d a mod d)%Z = (a mod d)%Z"
The 6th term has type "(Zmodd d a mod d)%Z = (a mod d)%Z" which should be coercible to
 "modEq ?d (Zmodd d a mod d) (a mod d)".
*)
rewrite Z.add_0_r; rewrite mod_mod. (* Produces the expected result. *)
```

/cc @Janno

#### Coq Version

8.15

**Comment by Janno** [comment]
It seems this might already be fixed on master but I need to re-check.

**Comment by Janno** [comment]
I think this is related to https://github.com/coq/coq/pull/15377 somehow but my `git bisect` abilities are insufficient to figure out what is going on here. In any case, I definitely cannot reproduce this on master.

**Comment by Janno** [comment]
A bisect between V8.14.0 and V8.15.0 to find the first time this bug occurrs points to
```
b0cf9adbf0ea9873f982993d15324c7328721fd1 is the first bad commit
commit b0cf9adbf0ea9873f982993d15324c7328721fd1
Author: Gaëtan Gilbert <gaetan.gilbert@skyskimmer.net>
Date:   Thu Dec 16 17:37:18 2021 +0100

    rewrite.ml: pass full "cstr" to transitivity
    
    Fix #10923
    
    (cherry picked from commit 53f9a6a31d34f33f09bcd9fb9c35127cf0afd9f8)
```

**Comment by Janno** [comment]
On master this was fixed by
```
0b82e46c5a935461f66ef59c367d7742b8a53e5e is the first bad commit
commit 0b82e46c5a935461f66ef59c367d7742b8a53e5e
Author: Matthieu Sozeau <mattam@mattam.org>
Date:   Fri Mar 19 17:14:48 2021 +0100

    Fix bug #7916, allowing `Reflexive` to have mode `! !`
```

**Comment by SkySkimmer** [comment]
That's not very backportable ;)
https://github.com/coq/coq/pull/15577 should be backportable though (or maybe merged straight to 8.15 branch?)
